pkg/autoscaler/aggregation/max/window.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package max
+
+import (
+	"fmt"
+
+	"github.com/davecgh/go-spew/spew"
+)
+
+type entry struct {
+	value float64
+	index int
+}
+
+// Window is a circular buffer which keeps track of the maximum value observed in a particular time.
+// Based on the "ascending minima algorithm" (http://web.archive.org/web/20120805114719/http://home.tiac.net/~cri/2001/slidingmin.html).
+type Window struct {
+	maxima        []entry
+	first, length int
+}
+
+// NewWindow creates an descending minima window buffer of size size.
+func NewWindow(size int) *Window {
+	return &Window{
+		maxima: make([]entry, size),
+	}
+}
+
+// Record records a value for a monotonically increasing index.
+func (m *Window) Record(index int, v float64) {
+	// Step One: Remove any elements where v > element.
+	// An element that's lower than the new element can never influence the
+	// maximum again, because the new element is both larger _and_ more
+	// recent than it.
+	for l := m.length - 1; l >= 0; l-- {
```

Comment:
we can go fancy to:

```
for ; v >=m.max[idx(m.length-1)]; m.length-- {}
```

Reply:
we do `m.length--` in the loop so won't work :(

Reply:
please no :joy:

Reply:
> we do m.length-- in the loop so won't work :(

eh, didn't follow — why not? :)

Reply:
😂😂😂

Reply:
CodeGolf as a credo

Reply:
@vagababov the loop decrements m.length to "remove" an element, so you could end up decrementing twice when you remove

Reply:
For posterity:
```
for ;m.length > 0 && v >= m.maxima[m.index(m.first+m.length-1)].value; m.length-- {}
```
a closer to actual code (may be even actual code) :)

Reply:
code above works and benchmarks are the same and it's pretty neat 🤔. I'm torn though about whether it sacrifices clarity or adds it vs. the super explicit but actually way more verbose than needed loop we have here. Gonna defer to @markusthoemmes on this and do whatever he says :).

Reply:
I already said "please no" :joy:

Reply:
But it's shows the versatility of go code :)